src/lib/tx/transaction.rs

```diff
@@ -0,0 +1,445 @@
+// Copyright (c) DUSK NETWORK. All rights reserved.
+// Licensed under the MPL 2.0 license. See LICENSE file in the project root for details.
+
+//! Phoenix transaction structure implementation.
+
+use crate::tx::{Crossover, Fee};
+use dusk_plonk::bls12_381::Scalar as BlsScalar;
+use dusk_plonk::proof_system::Proof;
+use phoenix_core::note::Note;
+use phoenix_core::Error;
+use std::convert::{TryFrom, TryInto};
+use std::io::{self, Read, Write};
+use std::mem::transmute_copy;
+
+/// Type identifiers for Phoenix transactions.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum TxType {
+    Transfer,
+    Distribute,
+    WithdrawFees,
+    Bid,
+    Stake,
+    Slash,
+    WithdrawStake,
+    WithdrawBid,
+}
+
+impl From<TxType> for u32 {
+    fn from(value: TxType) -> u32 {
+        match value {
+            TxType::Transfer => 0,
+            TxType::Distribute => 1,
+            TxType::WithdrawFees => 2,
+            TxType::Bid => 3,
+            TxType::Stake => 4,
+            TxType::Slash => 5,
+            TxType::WithdrawStake => 6,
+            TxType::WithdrawBid => 7,
+        }
+    }
+}
+
+impl TryFrom<u32> for TxType {
+    type Error = Error;
+
+    fn try_from(value: u32) -> Result<TxType, Error> {
+        match value {
+            0 => Ok(TxType::Transfer),
+            1 => Ok(TxType::Distribute),
+            2 => Ok(TxType::WithdrawFees),
+            3 => Ok(TxType::Bid),
+            4 => Ok(TxType::Stake),
+            5 => Ok(TxType::Slash),
+            6 => Ok(TxType::WithdrawStake),
+            7 => Ok(TxType::WithdrawBid),
+            n => Err(Error::InvalidNoteType(n as u8)),
+        }
+    }
+}
+
+impl Read for TxType {
+    fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
+        buf.write(&(u32::from(*self) as u8).to_le_bytes())
+    }
+}
+
+impl Write for TxType {
+    fn write(&mut self, mut buf: &[u8]) -> io::Result<usize> {
+        let mut n = 0;
+        let mut one_byte = [0u8; 1];
+
+        n += buf.read(&mut one_byte)?;
+        *self = (one_byte[0] as u32).try_into()?;
+
+        Ok(n)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
```

Comment:
Type has nothing to do with the contract that you're calling. 

Therefore, this should be removed. And Read and Write types might not need to be defined just for `type` since it is just an `u8`